pkg/lockfile/match-nuget-csproj.go

```diff
@@ -0,0 +1,159 @@
+package lockfile
+
+import (
+	"encoding/xml"
+	"errors"
+	"io"
+	"os"
+
+	"github.com/google/osv-scanner/internal/utility/fileposition"
+	"github.com/google/osv-scanner/pkg/models"
+
+	"path/filepath"
+	"strings"
+)
+
+// NugetCsprojMatcher matches the source file of a Nuget lockfile
+// https://learn.microsoft.com/en-us/nuget/consume-packages/package-references-in-project-files#locking-dependencies
+type NugetCsprojMatcher struct{}
+
+type NugetCsProj struct {
+	XMLName    xml.Name    `xml:"Project"`
+	ItemGroups []ItemGroup `xml:"ItemGroup"`
+}
+
+type ItemGroup struct {
+	XMLName           xml.Name           `xml:"ItemGroup"`
+	PackageReferences []PackageReference `xml:"PackageReference"`
+}
+
+type PackageReference struct {
+	XMLName     xml.Name `xml:"PackageReference"`
+	IncludeAttr *string  `xml:"Include,attr"`
+	Include     *string  `xml:"Include"`
+	VersionAttr *string  `xml:"Version,attr"`
+	Version     *string  `xml:"Version"`
+	models.FilePosition
+}
+
+func (itemGroup *ItemGroup) UnmarshalXML(decoder *xml.Decoder, start xml.StartElement) error {
+DecodingLoop:
+	for {
+		lineStart, columnStart := decoder.InputPos()
+		token, err := decoder.Token()
+		if err != nil {
+			return err
+		}
+		switch elem := token.(type) {
+		case xml.StartElement:
+			if elem.Name.Local != "PackageReference" {
+				continue
+			}
+
+			packageReference := PackageReference{}
+			packageReference.SetLineStart(lineStart)
+			packageReference.SetColumnStart(columnStart)
+			err := decoder.DecodeElement(&packageReference, &elem)
+			if err != nil {
+				return err
+			}
+			lineEnd, columnEnd := decoder.InputPos()
+			packageReference.SetLineEnd(lineEnd)
+			packageReference.SetColumnEnd(columnEnd)
+			itemGroup.PackageReferences = append(itemGroup.PackageReferences, packageReference)
+		case xml.EndElement:
+			if elem.Name == start.Name {
+				break DecodingLoop
+			}
+		}
+	}
+
+	return nil
+}
+
+func (m NugetCsprojMatcher) GetSourceFile(lockfile DepFile) (DepFile, error) {
+	var dir = filepath.Dir(lockfile.Path())
+
+	var dirs, err = os.ReadDir(dir)
+	if err != nil {
+		return nil, err
+	}
+
+	for _, file := range dirs {
+		if strings.HasSuffix(file.Name(), ".csproj") {
+			return OpenLocalDepFile(filepath.Join(dir, file.Name()))
+		}
+	}
```

Comment:
I believe we can have multiple csproj in the folder ? 
If yes, then shouldn't we scan all of them to try to find dependencies in the lockfile ?

Reply:
It's theoretically possible but not extensively used in practive.

Let's keep it as a limitation for the moment.